### PR TITLE
Optimize SHOW MEASUREMENTS so it consults the database index directly

### LINKED ISSUE #6950: Performance of SHOW MEASUREMENTS needs improvement
### Bug report

__System info:__ [Include InfluxDB version, operating system name, and other relevant details]

InfluxDB 0.13+

__Steps to reproduce:__

query `SHOW MEASUREMENTS`, it often takes multiple seconds.

@mjdesa is working up a test of response times vs. measurement count.

### PR DESCRIPTION
**Pull request by jsternberg**
SHOW MEASUREMENTS doesn't need to visit every shard in the open source
version since all of them contain the same database index.

Fixes #6950.

### CODE REVIEW

**Comment by jsternberg** [comment]
Performance comparisons:

```
> select * from "measurement-query-lat_show_measurements" group by xid
name: measurement-query-lat_show_measurements
tags: xid=egd61eatsv
time                            count   latS
----                            -----   ----
2016-07-08T20:17:11.374618Z     0       0.0437838698271662
2016-07-08T20:17:12.053288Z     10000   0.09904343215748668
2016-07-08T20:17:12.73149Z      20000   0.13075450202450156
2016-07-08T20:17:13.395655Z     30000   0.17315064719878137
2016-07-08T20:17:14.082572Z     40000   0.2354343249462545
2016-07-08T20:17:14.772045Z     50000   0.2925043071154505
2016-07-08T20:17:15.448546Z     60000   0.34408078785054386
2016-07-08T20:17:16.115443Z     70000   0.39206090895459056
2016-07-08T20:17:16.702316Z     80000   0.36979701300151646
2016-07-08T20:17:17.504219Z     90000   0.4069674031343311
2016-07-08T20:17:18.318817Z     100000  0.5777572281658649
2016-07-08T20:17:19.552159Z     110000  0.6584283879492432
2016-07-08T20:17:20.589572Z     120000  0.5762680179905146
2016-07-08T20:17:21.853695Z     130000  0.7345301820896566
2016-07-08T20:17:23.014017Z     140000  0.7709693089127541
2016-07-08T20:17:24.087094Z     150000  0.7063783928751945
2016-07-08T20:17:25.416668Z     160000  0.7431267169304192
2016-07-08T20:17:26.673182Z     170000  0.8820338801015168
2016-07-08T20:17:27.757723Z     180000  0.8194997408427298
2016-07-08T20:17:29.105029Z     190000  1.0249834167771041

name: measurement-query-lat_show_measurements
tags: xid=egd5zk4mzn
time                            count   latS
----                            -----   ----
2016-07-08T20:15:20.197796Z     0       0.03405444882810116
2016-07-08T20:15:20.858344Z     10000   0.06952855898998678
2016-07-08T20:15:21.541001Z     20000   0.11654633795842528
2016-07-08T20:15:22.157275Z     30000   0.13321433891542256
2016-07-08T20:15:22.821621Z     40000   0.1553444410674274
2016-07-08T20:15:23.520849Z     50000   0.18920857808552682
2016-07-08T20:15:24.234713Z     60000   0.23218040214851499
2016-07-08T20:15:24.863575Z     70000   0.2561171210836619
2016-07-08T20:15:25.54309Z      80000   0.28399864095263183
2016-07-08T20:15:26.24385Z      90000   0.38263382483273745
2016-07-08T20:15:26.866565Z     100000  0.3628534390591085
2016-07-08T20:15:27.705451Z     110000  0.3875560159794986
2016-07-08T20:15:28.470705Z     120000  0.48602429684251547
2016-07-08T20:15:29.492107Z     130000  0.4479611827991903
2016-07-08T20:15:30.312977Z     140000  0.5742275391239673
2016-07-08T20:15:31.342218Z     150000  0.5125550988595933
2016-07-08T20:15:32.488025Z     160000  0.5646843360736966
2016-07-08T20:15:33.870167Z     170000  0.6005091611295938
2016-07-08T20:15:35.014229Z     180000  0.626990145072341
2016-07-08T20:15:36.291056Z     190000  0.6528390720486641
```

There is likely more that can be done, but that seems like a decent improvement. I'll start working on formal benchmarks within the testing suite too.

**Comment by jwilder** [comment]
@benbjohnson Can you take a look?

**Comment by jsternberg** [comment]
I also have another potential change to add onto this that makes it a bit faster by keeping a sorted list of the measurements so we don't have to call `sort.Sort`. Should I push that too or keep it for a separate change?

The downside to that one is it causes a small hit to performance when creating or dropping a measurement.

**Comment by jwilder** [comment]
I would prefer a separate PR for that.

File: coordinator/statement_executor.go [hunk 1]
```diff
@@ -606,6 +517,137 @@ func (e *StatementExecutor) executeSelectStatement(stmt *influxql.SelectStatemen
...
+			// Apply limit & offset.
+			if opt.Limit > 0 || opt.Offset > 0 {
+				input = influxql.NewLimitIterator(input, opt)
+			}
+
+			aitr := influxql.NewAuxIterator(input, opt)
+			itr := aitr.Iterator("name", influxql.String)
+			aitr.Background()
+			return []influxql.Iterator{itr}, stmt, nil
+		}
```
Comment (benbjohnson) [inline coordinator/statement_executor.go]: This block instead the double nested `if` might be cleaner if pulled out to its own function (e.g. `createMeasurementsIterators()`. Then it won't be so indented and this function won't be quite so long.

**Comment by benbjohnson** [comment]
One comment about extracting block to a function but otherwise lgtm

File: coordinator/statement_executor.go (hunk 1 above)
Comment (jsternberg) [inline coordinator/statement_executor.go]: Do you mean extra the double nested `if` to its own function or extract the contents inside of the double nested `if` to another function?

Comment (benbjohnson) [inline coordinator/statement_executor.go]: Just the contents to another function. The function can accept the types that you asserted here.

Comment (jsternberg) [inline coordinator/statement_executor.go]: Ok, updated. Take another look to ensure I understood you correctly. Thanks.

File: coordinator/statement_executor.go [hunk 2]
```diff
@@ -606,6 +517,141 @@ func (e *StatementExecutor) executeSelectStatement(stmt *influxql.SelectStatemen
...
+func (e *StatementExecutor) createIterators(stmt *influxql.SelectStatement, ctx *influxql.ExecutionContext) ([]influxql.Iterator, *influxql.SelectStatement, error) {
+	// Handle SHOW MEASUREMENTS at the database level instead of delegating it to the shards.
+	if source, ok := stmt.Sources[0].(*influxql.Measurement); ok && source.Name == "_measurements" {
+		// Use the optimized version only if we have direct access to the database.
+		if store, ok := e.TSDBStore.(LocalTSDBStore); ok {
+			index := store.DatabaseIndex(source.Database)
+			if index == nil {
+				return nil, stmt, nil
+			}
+			itrs, err := e.createMeasurementsIterator(source.Database, store)
```
Comment (benbjohnson) [inline coordinator/statement_executor.go]: Yes, that's what I was thinking although this invocation passes two arguments and the function signature only accepts the `index`.

Comment (jsternberg) [inline coordinator/statement_executor.go]: Whoops. I should have compiled and ran the tests before pushing.

**Pull request merged**